lib/frrscript.c

```diff
@@ -183,61 +242,87 @@ void frrscript_register_type_codecs(struct frrscript_codec *codecs)
 		frrscript_register_type_codec(&codecs[i]);
 }
 
-struct frrscript *frrscript_load(const char *name,
-				 int (*load_cb)(struct frrscript *))
+struct frrscript *frrscript_new(const char *name)
 {
 	struct frrscript *fs = XCALLOC(MTYPE_SCRIPT, sizeof(struct frrscript));
 
 	fs->name = XSTRDUP(MTYPE_SCRIPT, name);
-	fs->L = luaL_newstate();
-	frrlua_export_logging(fs->L);
+	fs->lua_function_hash =
+		hash_create(lua_function_hash_key, lua_function_hash_cmp,
+			    "Lua function state hash");
+	return fs;
+}
+
+int frrscript_load(struct frrscript *fs, const char *function_name,
+		   int (*load_cb)(struct frrscript *))
+{
+
+	/* Set up the Lua script */
+	lua_State *L = luaL_newstate();
 
-	char fname[MAXPATHLEN * 2];
-	snprintf(fname, sizeof(fname), "%s/%s.lua", scriptdir, fs->name);
+	frrlua_export_logging(L);
 
-	int ret = luaL_loadfile(fs->L, fname);
+	char script_name[MAXPATHLEN * 2];
+
+	snprintf(script_name, sizeof(script_name), "%s/%s.lua", scriptdir,
+		 fs->name);
+	int ret = luaL_dofile(L, script_name);
 
 	switch (ret) {
 	case LUA_OK:
 		break;
 	case LUA_ERRSYNTAX:
-		zlog_err("Failed loading script '%s': syntax error: %s", fname,
-			 lua_tostring(fs->L, -1));
+		zlog_err(
+			"frrscript: failed loading script '%s.lua': syntax error: %s",
+			script_name, lua_tostring(L, -1));
 		break;
 	case LUA_ERRMEM:
-		zlog_err("Failed loading script '%s': out-of-memory error: %s",
-			 fname, lua_tostring(fs->L, -1));
+		zlog_err(
+			"frrscript: failed loading script '%s.lua': out-of-memory error: %s",
+			script_name, lua_tostring(L, -1));
 		break;
 	case LUA_ERRGCMM:
 		zlog_err(
-			"Failed loading script '%s': garbage collector error: %s",
-			fname, lua_tostring(fs->L, -1));
+			"frrscript: failed loading script '%s.lua': garbage collector error: %s",
+			script_name, lua_tostring(L, -1));
 		break;
 	case LUA_ERRFILE:
-		zlog_err("Failed loading script '%s': file read error: %s",
-			 fname, lua_tostring(fs->L, -1));
+		zlog_err(
+			"frrscript: failed loading script '%s.lua': file read error: %s",
+			script_name, lua_tostring(L, -1));
 		break;
 	default:
-		zlog_err("Failed loading script '%s': unknown error: %s", fname,
-			 lua_tostring(fs->L, -1));
+		zlog_err(
+			"frrscript: failed loading script '%s.lua': unknown error: %s",
+			script_name, lua_tostring(L, -1));
 		break;
 	}
 
 	if (ret != LUA_OK)
 		goto fail;
 
+	/* Push the Lua function we want */
+	lua_getglobal(L, function_name);
+	if (lua_isfunction(L, lua_gettop(L)) == 0)
+		goto fail;
+
 	if (load_cb && (*load_cb)(fs) != 0)
 		goto fail;
```

Comment:
shouldn't these two above be logging an error on failure?

Reply:
updated in https://github.com/FRRouting/frr/pull/8982/commits/4093300ee891bd7559b93e2b7c94fb6fc68b5ddc; yup these should be error until we have knobs to express which hooks calls should always run vs run if present